### ISSUE TITLE
Fine control over upload resources

### ISSUE THREAD

**Issue by joiskash** [google/android-fhir#1483]
**Is your feature request related to a problem? Please describe.**
The download manager allows you to pick and choose what you want to download.
The uploader creates a bundle with all the local changes.  We would like to have control over which resources get uploaded. Some of the use cases could be :

1. Uploading a patient resource + the references related to it .
2. Breaking up the bulk upload bundle and limiting it to a certain size.

Both these use case are related to the offline use of the app. We have noticed that a 1555477 Bytes bundle takes around 44341ms. This includes 50 Patient resources and 10 to 15 resources per patient. 
So if this doubles it will be very close to the server time out. Which will result in sync failures. Any subsequent sync will fail.

**Describe the solution you'd like**
Have more control over the resources that are being uploaded.

cc: @jingtang10

**Comment by jingtang10** [comment]
cc @omarismail94 @aditya-07

**Comment by Itskiprotich** [comment]
Similar to this, I have a scenario where I'm pulling resources. I.e Encounters referencing other encounters using the Encounter.partOf. The challenge comes in when only the observations for the parent Encounter are pulled while the other Encounters have blank observations. Any workaround on this? Thanks

**Comment by kumaakh** [comment]
@jjtswan

**Comment by joiskash** [comment]
Update on sync - 

- We have an app running fhir engine with a feature to create fake patients with fake resources. Each fake patient creation, through pre filled questionnaires and extraction results in one Patient resource and 46 other related resources.

- We came across an OOM while uploading many unsynced resources . Approx 300 patients = 300 * 46 = 13800 FHIR resources. The OOM occurs when parsing json payload.

- We modified upload to group together 10 resources at a time for upload.
